torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1351,62 +1364,66 @@ For* LoopNest::fuseLoops(const std::vector<For*>& loops) {
     return diff->isConstant() && (immediateAs<int>(diff) == 0);
   };
+  auto first_loop = loops.front();
   auto first_loop_start = IRSimplifier::simplify(first_loop->start());
   auto first_loop_stop = IRSimplifier::simplify(first_loop->stop());
   for (size_t i = 1; i < loops.size(); ++i) {
     auto curr_loop = loops[i];
     auto curr_loop_start = IRSimplifier::simplify(curr_loop->start());
     auto curr_loop_stop = IRSimplifier::simplify(curr_loop->stop());
     if (!are_bounds_equal(curr_loop_start, first_loop_start)) {
-      throw malformed_input(
-          "Loops with different start bounds cannot be fused");
+      return false;
     }
     if (!are_bounds_equal(curr_loop_stop, first_loop_stop)) {
-      throw malformed_input("Loops with different stop bounds cannot be fused");
+      return false;
     }
   }
 
-  // Now, we fuse the incoming loops. This is done by taking all the statements
-  // from the second loops onwards and moving them into the first loop's body.
-  // This way the final fused loop will be the same as the first loop.
-  for (size_t i = 1; i < loops.size(); ++i) {
-    auto body = dynamic_cast<Block*>(Substitute(
-        Stmt::clone(loops[i]->body()), {{loops[i]->var(), first_loop->var()}}));
-    first_loop->body()->splice(first_loop->body()->end(), body);
-    root_block->remove_stmt(loops[i]);
-  }
-
-  analysis::MemDependencyChecker analyzer;
-  first_loop->accept(&analyzer);
-  for (auto it1 = first_loop->body()->begin(); it1 != first_loop->body()->end();
-       ++it1) {
-    for (auto it2 = std::next(it1); it2 != first_loop->body()->end(); ++it2) {
-      if (hasPartialOverlap(analyzer, *it2, *it1)) {
-        // Whenever there is a partial overlap between accesses in 2 statements
-        // in a loop, it results in a loop-carried dependence. NOTE: In
-        // general, this may not be true. But the semantics of TE IR is that all
-        // iterations of the loop can run in parallel, so there should be no
-        // loop-carried dependence in either direction (forward or backward).
-        throw malformed_input(
-            "Fusing given loops is not valid since it results in a loop carried dependence");
-      }
+  // A lambda to fuse all the given loops.
+  auto fuse_all_loops = [](const std::vector<For*>& loops) {
+    auto first_loop = loops.front();
+    // Fuse the loops by taking all the statements from the second loops
+    // onwards and moving them into the first loop's body.
+    // This way the final fused loop will be the same as the first loop.
+    for (size_t i = 1; i < loops.size(); ++i) {
+      auto body = dynamic_cast<Block*>(Substitute(
+          Stmt::clone(loops[i]->body()),
+          {{loops[i]->var(), first_loop->var()}}));
+      first_loop->body()->splice(first_loop->body()->end(), body);
     }
+  };
+
+  // We need to check if fusing the loops results in a loop-carried dependence.
+  // This check can be done only after the loops are fused into one. But if the
```

Comment:
Really?  It seems kind of sketchy that we need to mutate the IR to check if mutating the IR is valid.

Reply:
Here is the relevant safety discussion from the paper that introduced loop fusion (I think):

> In the case of a loop-independent dependence, fusion is safe if the sense of
the dependence is preserved, i.e., if the dependence direction is not reversed. A
simple test for this case performs dependence testing on the loop bodies as if they
were in a single loop. After fusion, a loop-independent dependence between the
original nests can (a) remain loop-independent, (b) become forward loop-carried
or (c) become backward loop-carried. Since the direction of the dependence is
preserved in the rst two cases, fusion is legal. Fusion is illegal when a loopindependent dependence becomes a backward carried dependence after fusion.
These dependences are called fusion-preventing dependences [1, 25].

More about this on page 3 here: https://www.cs.utexas.edu/users/mckinley/papers/lcpc-1993.pdf

So, the check has to be done as if the fusion already happened. We can try to figure out a way to do this without fusing but this felt more intuitive and cleaner to me.

Reply:
Yeah, I mean I get that the dependence test is done "as if" it were a single loop, but actually mutating IR feels like a pretty wasteful way of implementing "as if".  In any case, I'll accept to unblock and we can optimize this later.

Reply:
I thought about the "as if" part. We will at least have to clone the individual statements and replace the loop variables to match all statements. The only additional thing the fusion step here is doing is to bring them all under one body. May be I can avoid that. Will do that later.

Reply:
Yeah, if we've got to clone and replace the induction variables we're already kinda screwed.  Don't worry about it.